File: observer_pattern/sese/WeatherData.py
from Interface import Observer

# 주제인 WeatherData 구현
class WeatherData():
    
    # 옵저버를 저장하는 list
    def __init__(self):
        self._observers = []

    # 옵저버 등록
    def registerObserver(self, Observer):
        observer = Observer()
        self._observers.append(observer)

    # 옵저버 제거
    def removeObserver(self, Observer):
        observer = Observer()
        self._observers.remove(observer)
    
    # 옵저버에게 알림
    def notifyObservers(self):
        for observer in self._observers:
            observer.update(self)

    # 측정값 변화 감지
    def measurementsChanged(self):
        self.notifyObservers()

    # 측정값 설정
    def setMeasurements(temperature, humidity, pressure):
        self.measurementsChanged()

File: observer_pattern/hasun/WeatherORama/WeatherData.java
package WeatherORama;

import java.util.ArrayList;
import java.util.List;

public class WeatherData implements Subject {
    private List<Observer> observers;   // Observer 객체들을 저장하는 ArrayList
    private float temperature;
    private float humidity;
    private float pressure;

    public WeatherData() {
        observers = new ArrayList<Observer>();
    }

    @Override
    public void registerObserver(Observer o) {
        observers.add(o);
    }

    @Override
    public void removeObserver(Observer o) {
        observers.remove(o);
    }

    @Override
    public void notifyObserver() {
        for (Observer observer : observers) {
            observer.update();
        }
    }

    public void measurementsChanged() {
        notifyObserver();
    }

    public void setMeasurements(float temperature, float humidity, float pressure) {
        this.temperature = temperature;
        this.humidity = humidity;
        this.pressure = pressure;
        measurementsChanged();
    }

    public float getPressure() {
        return this.pressure;
    }

    public float getHumidity() {
        return this.humidity;
    }

    public float getTemperature() {
        return this.temperature;
    }
}
